File: Data Structures & Algorithms/shortest-path-in-binary-matrix/submission-1.py
from collections import deque

class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])

        if grid[0][0] or grid[ROWS - 1][COLS - 1]:
            return -1

        queue = deque([(0,0,1)])
        visited = set((0,0))
        dirs = [(-1 , 0), (0, -1), (1, 0), (0, 1), (-1, -1), (1,1), (-1, 1), (1, -1)]

        while queue:
            for i in range(len(queue)):
                r,c, length = queue.popleft()
                if r == ROWS - 1 and c == COLS - 1:
                    return length
                
                for dr, dc in dirs:
                    row, col = r + dr, c + dc
                    if min(row, col) < 0 or row == ROWS or col == COLS or (row, col) in visited or grid[row][col] == 1:
                        continue
                    queue.append((row, col, length + 1))
                    visited.add((row, col))
        
        return -1 

File: Data Structures & Algorithms/rotting-fruit/submission-1.kt
class Solution {
    fun orangesRotting(grid: Array<IntArray>): Int {
        val queue = ArrayDeque<Pair<Int, Int>>()
        val directions = arrayOf(
            intArrayOf(0,-1),
            intArrayOf(0,1),
            intArrayOf(-1,0),
            intArrayOf(1,0),
        )
        val rows : Int = grid.size
        val cols : Int = grid[0].size

        var fresh : Int = 0
        var time : Int = 0

        for(r in 0 until rows){
            for(c in 0 until cols){
                if(grid[r][c] == 2){
                    queue.addLast(r to c)
                }
                if(grid[r][c] == 1){
                    fresh++
                }
            }
        }

        while(queue.isNotEmpty() && fresh > 0){
            for(i in 0 until queue.size){
                val (row, col) = queue.removeFirst()
                for(dir in directions){
                    val nr = row + dir[0]
                    val nc = col + dir[1]
                    if(nr !in 0 until rows || nc !in 0 until cols || grid[nr][nc] != 1){
                        continue
                    }
                    queue.addLast(nr to nc)
                    grid[nr][nc] = 2
                    fresh--
                }
            }
            time++
        }

        return if(fresh == 0) time else -1

    }
}
